Welcome Plugin

adamharrison opened the pull request:
Adds in a welcome plugin that offers to download certain metapackages on first startup.

Guldoman commented on plugins/plugin_manager/init.lua:
```diff
@@ -359,4 +366,9 @@ if pcall(require, "plugins.terminal") then
   })
 end
 
+keymap.add({
+  ['ctrl+shift+1'] = 'plugin-manager:show',
+  ['ctrl+shift+2'] = 'plugin-manager:open-session'
+})
+
```
Added for testing?

adamharrison replied:
Nah; should actually be a change on master (but whatever, may as well add it here; gives an easy way to open the plugin manager. Think I should nix it?

Guldoman replied:
Yep

Guldoman commented on plugins/welcome.lua:
```diff
@@ -0,0 +1,124 @@
+  end
+})
+
+return {  }
```
Should probably just not return anything.

adamharrison replied:
OK.

Guldoman commented on plugins/welcome.lua:
```diff
@@ -0,0 +1,124 @@
+local PluginManager = require "plugins.plugin_manager"
+local PluginView = require "plugins.plugin_manager.plugin_view"
+
+local welcomed = system.get_file_info(USERDIR .. PATHSEP .. "welcomed") ~= nil
```
I wonder if you should just return here, if `welcomed` is true.
You could even do it as the first thing this plugin does, even before the `require`s.

adamharrison replied:
True! Will do that.

Guldoman commented on plugins/welcome.lua:
```diff
@@ -0,0 +1,124 @@
+  { label = "Install Addons Package", command = "welcome:install-addons", tooltip = {
+    "Will install the basic addons package for lite-xl.",
+    "",
+    "Includes all syntax higlightings, various themes, as well as a few extra plugins to help lite-xl a bit more easy to interact with.",
```
```suggestion
    "Includes all syntax highlightings, various themes, as well as a few extra plugins to help lite-xl be a bit more easy to interact with.",
```

Guldoman commented on plugins/welcome.lua:
```diff
@@ -0,0 +1,124 @@
+      y = y + style.font:get_height()
+    end
+  else
+    common.draw_text(style.font, style.dim, "Hover over one of the options below to get started.", "center", self.position.x, y, self.size.x, style.font:get_height())
```
```suggestion
    common.draw_text(style.font, style.dim, "Hover over one of the options above to get started.", "center", self.position.x, y, self.size.x, style.font:get_height())
```

Guldoman reviewed (commented):
All mentions of `lite-xl` should be replaced by `Lite XL`.
The "tooltip" text is too dim. Should probably use the same color as the "entries" text.
The buttons aren't interactive.
Opening the plugin manager dismisses the welcome screen.
Resizing the window should wrap the tooltips.
Resizing the window to a small enough size leads to the underlying `EmptyView` to show up.

I'm still not sold on the overall design and UX, but this will do well enough for now.

adamharrison commented:
> All mentions of lite-xl should be replaced by Lite XL.
> The "tooltip" text is too dim. Should probably use the same color as the "entries" text.

Fair.

> The buttons aren't interactive.

lol; nice catch. Will fix. 

> Opening the plugin manager dismisses the welcome screen.

By design; do you think we should keep the welcome screen around even after they've chosen to open the manager?

> Resizing the window should wrap the tooltips.
> Resizing the window to a small enough size leads to the underlying EmptyView to show up.

Ugh. Will fix.

> I'm still not sold on the overall design and UX, but this will do well enough for now.

Yeah it's not intended to be the be all end all; please feel free to suggest other designs.

Guldoman commented:
> By design; do you think we should keep the welcome screen around even after they've chosen to open the manager?

I feel like it doesn't imply enough that clicking that will dismiss the welcome screen.

Btw, I'm also not sure about the decision of overriding `EmptyView`, but I guess it's fine for now.

adamharrison commented:
OK; I've changed things around a bit.

Linewrapping is hard; so instead I've simplified the text (removing extraneous syntax to make things clear, simple and unambiguous), which also reduces the width. I've also hidden the treeview during the welcome process, which I think is reasonable; so should be a lot harder to trigger a situation where text is offscreen.

I've also made a point to note situations where a network connection is required.

Thoughts?

Guldoman commented on plugins/welcome.lua:
```diff
@@ -12,51 +12,84 @@ local Node = require "core.node"
+    xv = x1 - (version_width - title_width)
+  end
+  x = renderer.draw_text(style.big_font, title, x1, y1, style.dim)
+  renderer.draw_text(style.font, version, xv, y1 + th, style.dim)
```
`xv` can be uninitialized here.

adamharrison replied:
Fixed.
